Refactor: Editor separation from the core engine

miogds opened the pull request:
This PR introduces a major architectural refactor to better separate **core engine runtime functionality** from **editor-specific tooling**. The changes improve modularity, reduce coupling, and establish a cleaner foundation for future editor and system work.  

The focus of this PR is on the **minimal set of changes required** to have a **working editor** without unnecessary complexity. The goal was to make the transition as smooth as possible, remove only what was strictly necessary, and ensure everything compiles cleanly. While there is still plenty of room for improvement, this provides a **stable baseline** with a **clear path forward** for future features and refinements.  

## Key Changes  

### Architectural Restructuring  
- Moved editor-specific components into a new **`UntoldEngineEditor`** module.  
- Extracted editor views, systems, and utilities into a dedicated module.  
- Reorganized rendering pipelines into a dedicated **`Pipelines`** directory.  
- Removed editor dependencies (e.g., **HotReloading**, **VisualDebugger**) from the **core engine**.  
- Renamed **`Camera.swift` → `CameraSystem.swift`** for naming consistency.  
- Core engine now focuses exclusively on runtime systems, while the editor manages tooling.  

### Core Engine Updates  
- **Input System**: Added `InputSystem+Keyboard.swift` extension to handle keyboard input separately.  
- **Camera System**: Simplified to focus on runtime-only behavior (removed editor-specific logic).  
- **Render Passes**: Refactored pipeline to be modular and independent of the editor.  
- **Global State**: Updated to support distinct runtime vs. editor contexts.  

### Editor Updates  
- **Camera System**: Added dedicated editor camera system (`CameraSystem.swift`) with editor-only controls.  
- **Input System**: Removed old `InputSystem.swift`; introduced AppKit-based input handling for editor use.  
- **Renderer**: Updated to run independently from the core engine renderer.  
- **Editor Views**: Modified views and controllers to integrate with decoupled systems.  

### Game Project Updates  
- **DemoGame & StarterGame**: Updated to work with the new modular architecture.  
- **Systems**: Adjusted ball, dribbling, and camera follow systems for compatibility.

untoldengine commented:
Hi Javier,
Thank you for the PR. For some reason, it is not building for me. Here is the error output. I also noticed that it failed in the CI build as well. 

`Renderer/UntoldEngine.swift:228:13: error: call to main actor-isolated instance method 'initResources()' in a synchronous nonisolated context
    initResources()
    ^

Renderer/UntoldEngine.swift:94:17: note: calls to instance method 'initResources()' from outside of its actor context are implicitly asynchronous
    public func initResources() {
                ^

Renderer/UntoldEngine.swift:208:17: note: add '@MainActor' to make instance method 'mtkView(_:drawableSizeWillChange:)' part of global actor 'MainActor'
    public func mtkView(_ mtkView: MTKView, drawableSizeWillChange size: CGSize) {
                ^
    @MainActor

Scenes/SceneView.swift:25:24: error: call to main actor-isolated instance method 'initResources()' in a synchronous nonisolated context
    self.renderer?.initResources()
                   ^

Renderer/UntoldEngine.swift:94:17: note: calls to instance method 'initResources()' from outside of its actor context are implicitly asynchronous
    public func initResources() {
                ^

error: fatalError
`

untoldengine commented:
Hi Javier, thanks for this great effort. I'm still reviewing the code.  So far, I'm liking what I'm seeing. 

I noticed duplicate code in UpdateRenderingSystem() and EditorUpdateRenderingSystem().
Both functions build the command buffer, fetch the render pass descriptor, sort/execute the graph, present/commit, and run the same completion handler. Only the policy differs (which passes to add and which graph builder to call).

Why don't we remove the duplication by centralizing the frame “mechanics” in core, and pass the editor/game “policy” in. For example:

```
// Core (UntoldEngine) — one shared runner
public struct RenderGraphPolicy {
    let buildPrecompute: () -> ([String: RenderPass], String)
    let finalize: (_ graph: inout [String: RenderPass], _ after: String) -> Void
}

@MainActor
public func runFrame(in view: MTKView, policy: RenderGraphPolicy) {
    guard let cb = renderInfo.commandQueue.makeCommandBuffer() else { return }
    cb.label = "Rendering Command Buffer"

    executeFrustumCulling(cb)

    guard let rpd = view.currentRenderPassDescriptor else { cb.commit(); return }
    renderInfo.renderPassDescriptor = rpd

    var (graph, preID) = policy.buildPrecompute()
    policy.finalize(&graph, preID)

    do {
        let sorted = try topologicalSortGraph(graph: graph)
        executeGraph(graph, sorted, cb)
    } catch {
        print("Render graph sort failed: \(error)")
    }

    if let drawable = view.currentDrawable { cb.present(drawable) }
    cb.addCompletedHandler { _ in
        DispatchQueue.main.async {
            needsFinalizeDestroys = true
            visibleEntityIds = tripleVisibleEntities.snapshotForRead(frame: cullFrameIndex)
        }
    }
    cb.commit()
}
```

Then each target just supplies its policy and calls the shared runner:

```
// Game/app target
@MainActor
func UpdateRenderingSystem(in view: MTKView) {
    let policy = RenderGraphPolicy(
        buildPrecompute: { buildGameModeGraph() },
        finalize: { graph, id in
            graph["composite"] = RenderPass(id: "composite",
                                            dependencies: [id],
                                            execute: RenderPasses.compositeExecution)
        }
    )
    runFrame(in: view, policy: policy)
}

// Editor target
@MainActor
func EditorUpdateRenderingSystem(in view: MTKView, visualDebug: Bool) {
    let policy = RenderGraphPolicy(
        buildPrecompute: { buildEditModeGraph() },
        finalize: { graph, id in
            if visualDebug {
                graph["debug"] = RenderPass(id: "debug",
                                            dependencies: [id],
                                            execute: RenderPasses.debuggerExecution)
            } else {
                graph["composite"] = RenderPass(id: "composite",
                                                dependencies: [id],
                                                execute: RenderPasses.compositeExecution)
            }
        }
    )
    runFrame(in: view, policy: policy)
}

```

This keeps editor/core separation (core knows nothing about editor), removes duplicate render-loop code, and makes it easy to add more modes later (benchmark, capture, headless) by supplying a new policy.

miogds commented:
I just removed the MainActor property. This was a test for Swift 6 compilation and on my computer it works perfectly fine. Anyway, it can't be compiled in Swift 6 for now, as more changes are required, so I just removed it. We don't need it for now.

From your comments, I thought about not repeating the code, but I chose to repeat it in this PR because it was the fastest way to split the editor and leave the build graph refactoring for next PRs because I want to remove some of the global functions and tie them more to the renderer itself, so we can have different render pipelines and steps even per view. I want to study deeply on how I want to address that, so I took the faster path for now. 

I will propose you several options I was thinking about it, and I will need your help to better understand which could be the best way to do it. During this week, I will write several more discussions for the next PRs about the next steps for editor / engine separation.

untoldengine commented:
Ok, so I'm going to try to finish reviewing your PR by tomorrow evening. Again, I'm liking what I'm seeing. One thing, though, the tests are failing. It is complaining about the getMainCamera() missing and about the input system. Can't merge until tests are passing.
 
Also, with your changes, I think that we may not need DemoGame anymore since users can just select the editor target. I'll remove it after your PR is merged. 

Thanks.

untoldengine commented on Sources/UntoldEngineEditor/Systems/EditorCameraSystem.swift:
Can we change the name to EditorCameraSystem ? There is another CameraSystem under the core renderer. That way is clear which camera we are working with.

miogds replied:
Sure. I will change the name also I will change the test to be able to pass in swift test.

miogds replied:
I reviewed the code and the file you mention `Sources/UntoldEngineEditor/Systems/CameraSystem.swift`. I can rename it, but I did not because I follow the pattern of creating the same file and folder structure like the engine core to expand the functionality in the editor, at least for the first / fast approach. 

That's why I choose the same name as the core, and it's easy to go to the counterpart, the same file in the engine core. Anyway, if you still want me to change it, there are no issues with that. I'll change it?

untoldengine replied:
Yes, please. It will make it easier for future contributors to know where to add camera functionality related to the game and editor. Thank you

untoldengine replied:
I noticed the UntoldEngineEditor/Renderer folder has files with the same names as in the core renderer (e.g. both define RenderPasses.swift). To avoid confusion in searches and diffs, could we rename the editor versions with a clear suffix?

For example:
EditorRenderPasses.swift

This will make it obvious which module a file belongs to and help future contributors (and ourselves) avoid confusion when navigating the codebase.

Thank you

miogds replied:
I'll push the changes

untoldengine commented:
Hi,
Tests are still failing. I was able to point out the issue in some of the unit tests.
In the SteeringSystemTest.swift, line 494, it should be `InputSystem.shared.keyState.dPressed = true`

In CullingTest.swift, let's get rid of `var sceneCamera = EntityID!`. We should also remove lines 53-55

```
sceneCamera = createEntity()
createSceneCamera(entityId: sceneCamera)
createGameCamera(entityId: camera)
```

and since getMainCamera() was removed, line 154 should call `findGameCamera()`

In CameraTest.swift, references to `sceneCamera` should also be removed.

Also, currently looking into why the RendererTest.swift are failing. The main issue is that `func setUp()` should now be modified to follow the recent changes.

miogds commented:
Sorry about the tests, but they are not working on my computer, and Xcode does not pass all the tests that Swift tests do. So I had to upload, check the error, and test again. I will later try to figure out why they are not working on my Mac.

miogds commented:
It seems that the PR finally passed the tests.

untoldengine commented:
Thank you.  Going to test them again. 

Also, since this is a pretty big PR, I’m going to turn on GitHub Copilot code review as an extra set of eyes. Don’t worry — it’s not an enforcement step. Copilot just leaves automated suggestions that might catch small things I could miss. I’ll still be doing the full review myself, but this way we might also spot a few low-level issues or cleanup opportunities more quickly.

miogds commented on Sources/UntoldEngineEditor/Systems/HotReloadingSystem.swift:
```diff
@@ -125,8 +127,15 @@ func reloadPipeline(named pipelineName: String, with library: MTLLibrary, pipe:
+        //TODO: Check if the value actually changes or becasue its a struct we are just copy by value and it's not changing at all so we had to re-assign the pipe to the manager
         pipe.pipelineState = newPipeline
         pipe.depthState = newDepthState
+        
```
this one is a good catch! I forget to review this todo in the PR.

miogds commented on Sources/UntoldEngine/Renderer/Pipelines/ComputePipelines.swift:
```diff
@@ -0,0 +1,46 @@
+
+import MetalKit
+
+public struct ComputePipeline
+{
```
For the brackets open/close position, I have my particular rules in which cases it's better to put on line below or just after the end of the line, but this will be on the Syntax discussion in the future :)

untoldengine replied:
Don't worry about this. To be honest, I don't like to nitpick on stuff like this. Rather, focus on the big changes. I'm setting up instructions to Copilot not to be a nitpicker. :)

untoldengine commented:
@miogds sorry Javier, didn't expect Copilot to be so nitpicking LOL. You can ignore those nitpicking suggestions since swiftformat should fix all of that.

untoldengine commented:
Hi, I’m hitting this error when linking the engine as an SPM dependency. I also see the same error while running the Rendering Unit Tests:

`Error: 1006: Pipeline State is null for iblPreFilterPipeline is nil`

What I’m doing (high level):

- New macOS game project.
- Added UntoldEngine via SPM.
- Created a workspace and added both the game project and the engine as a local package.

This is essentially the boiler-plate code that I'm using:

```
import MetalKit
import SwiftUI
import UntoldEngine

// GameScene is where you would initialize your game and write the game logic.
class GameScene {
    init() {
        
    }

    func update(deltaTime _: Float) {
       
    }

    func handleInput() {
      
        // Handle input here
    }
}

// AppDelegate: Boiler plate code -- Handles everything – Renderer, Metal setup, and GameScene initialization
class AppDelegate: NSObject, NSApplicationDelegate {
    var window: NSWindow!
    var renderer: UntoldRenderer!
    var gameScene: GameScene!

    func applicationDidFinishLaunching(_: Notification) {
        print("Launching Untold Engine v0.2")

        // Step 1. Create and configure the window
        window = NSWindow(
            contentRect: NSRect(x: 0, y: 0, width: 1920, height: 1080),
            styleMask: [.titled, .closable, .resizable],
            backing: .buffered,
            defer: false
        )

        window.title = "Untold Engine v0.2"
        window.center()

        // Step 2. Initialize the renderer and connect metal content
        guard let renderer = UntoldRenderer.create() else {
            print("Failed to initialize the renderer.")
            return
        }

        window.contentView = renderer.metalView

        self.renderer = renderer

        renderer.initResources()

        // Step 3. Create the game scene and connect callbacks
        gameScene = GameScene()
        renderer.setupCallbacks(
            gameUpdate: { [weak self] deltaTime in self?.gameScene.update(deltaTime: deltaTime) },
            handleInput: { [weak self] in self?.gameScene.handleInput() }
        )

        let hostingView = NSHostingView(rootView: SceneView(renderer: renderer))
        window.contentView = hostingView

        window.makeKeyAndOrderFront(nil)
        NSApp.setActivationPolicy(.regular)
        NSApp.activate(ignoringOtherApps: true)
    }

    func applicationShouldTerminateAfterLastWindowClosed(_: NSApplication) -> Bool {
        true
    }
}
```

- In the test setup, I create an NSWindow, attach the MTKView, call initResources(), then render and read the offscreen targets.

Debugging this issue shows that the pipelines are being used before they exist (first MTKView callback runs before initResources()). I ended up doing a quick-and-dirty change where I modified the mtkview to call `initResources` instead of `initSizeableResources` and then removed the initResources() call from the test setup, and it worked fine.

untoldengine commented:
Hi Javier,

Quick question — when you set up an external project with the UntoldEngine package (via SPM), are you able to see the UntoldEngineEditor scheme in Xcode?

I followed the steps (created a workspace, added my game project, then added the engine as a local Swift package). I can link the UntoldEngine library just fine, but when I check Product → Scheme → Manage Schemes, I only see my game’s scheme — the editor doesn’t appear.

Can you confirm if you see the UntoldEngineEditor (and the other executables) on your side, or does Xcode only expose the library scheme when using the engine as a package?

Thanks!


**UPDATE** Never mind,  I was able to fix it. Had to add the editor as an executable and then restart Xcode for it to show up.

miogds commented:
That's weird, this compiles perfectly fine on my computer. Also, when I’m working on the Xcode workspace, I don't add the package through the package dependency fo Xcode. I just link from the frameworks linked page inside the general tab of the target executable.

miogds commented:
Ok. I see the error. 

I think the initResources should not be public, also the init function from the renderer could not be public as the create static function is the one responsible for creation that should call the init resources function inside to avoid this problem. 

I already pushed a fix. 

In the code about, you would remove the initResources(). But now the compiler will complaint about is internal now. 

I like more the approach where the developer just creates the renderer and the renderer does all the magic to avoid the developer having to understand internally what had to do to render something.

untoldengine commented:
ok, let me test it again. thanks

untoldengine commented on Sources/UntoldEngine/Renderer/UntoldEngine.swift:
```diff
@@ -77,35 +94,16 @@ public class UntoldRenderer: NSObject, MTKViewDelegate {
-        let sceneCamera = createEntity()
-        createSceneCamera(entityId: sceneCamera)
-
-        let gameCamera = createEntity()
```
Any reason why the game camera (not scene camera) and light creation were removed? Can we put them back, please? The game developer will forget to do this in their game. 

This will do

```
        let gameCamera = createEntity()
        setEntityName(entityId: gameCamera, name: "Main Camera")
        createGameCamera(entityId: gameCamera)

        let light = createEntity()
        setEntityName(entityId: light, name: "Directional Light")
        createDirLight(entityId: light)
```

I think that for the camera entity, you will also need to do `CameraSystem.shared.activeCamera = gameCamera` as per your changes.

miogds replied:
Well, I removed because I did not want to add objects that you will delete on loading a external scene file, because now you have to destroy the ones you create just seconds ago... 

Anyhow, you are right, the developer will forget, so I will figure in the future a better initialization process or a default configuration if the developer don't add the default ones... 

Done!. I Pushed the changes!

untoldengine commented:
Hi @miogds, thank you for the PR. It is great. I think I'm pretty much done with the review. Once you push the new commit with my suggestion, I'll approve it. 
After the merge, I'm going to have to make some minor changes to the unit test so that it follows your changes.

untoldengine commented:
PR has been approved. Thank you so much for this.

untoldengine commented:
Merging your branch.
